src/diff.rs

```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+// The parent must be on the stack when we enter this function:
+//
+//     [... parent]
+//
+// When this function exits, the stack remains the same.
+fn diff_keyed_suffix(
+    cached_set: &CachedSet,
+    change_list: &mut ChangeListBuilder,
+    registry: &mut EventsRegistry,
+    old: &[Node],
+    new: &[Node],
+    cached_roots: &mut FxHashSet<CacheId>,
+    new_shared_suffix_start: usize,
+) {
+    debug_assert_eq!(old.len(), new.len());
+    debug_assert!(!old.is_empty());
+
+    // [... parent]
+    change_list.push_child(new_shared_suffix_start);
+    // [... parent new_child]
+
+    for (old_child, new_child) in old.iter().zip(new.iter()) {
+        diff(
+            cached_set,
+            change_list,
+            registry,
+            old_child,
+            new_child,
+            cached_roots,
+        );
+
+        // [... parent this_new_child]
+        change_list.pop_push_next_sibling();
+        // [... parent next_new_child]
+    }
+
+    // [... parent]
+    change_list.pop();
+}
+
+// Diff children that are not keyed.
+//
+// The parent must be on the top of the stack when entering this function:
+//
+//     [... parent]
+//
+// The stack is in the same state when this function returns.
+fn diff_non_keyed_children(
+    cached_set: &CachedSet,
+    change_list: &mut ChangeListBuilder,
+    registry: &mut EventsRegistry,
+    old: &[Node],
+    new: &[Node],
+    cached_roots: &mut FxHashSet<CacheId>,
+) {
+    // Handled these cases in `diff_children` before calling this function.
+    debug_assert!(!new.is_empty());
+    debug_assert!(!old.is_empty());
+
+    for (i, (new_child, old_child)) in new.iter().zip(old.iter()).enumerate() {
+        if i == 0 {
             change_list.push_first_child();
+        } else {
+            change_list.pop_push_next_sibling();
         }
-        change_list.remove_self_and_next_siblings();
-        pushed = false;
-    } else {
-        debug!("  creating new children");
-        for (i, new_child) in new_children.enumerate() {
-            if i == 0 && pushed {
-                change_list.pop();
-                pushed = false;
-            }
-            create(
+
+        diff(
+            cached_set,
+            change_list,
+            registry,
+            old_child,
+            new_child,
+            cached_roots,
+        );
+    }
+
+    match old.len().cmp(&new.len()) {
+        Ordering::Greater => {
+            change_list.pop_push_next_sibling();
+            remove_self_and_next_siblings(change_list, registry, &old[new.len()..]);
+        }
+        Ordering::Less => {
+            change_list.pop();
+            create_and_append_children(
                 cached_set,
                 change_list,
                 registry,
-                new_child.clone(),
+                &new[old.len()..],
                 cached_roots,
             );
-            change_list.append_child();
+        }
+        Ordering::Equal => {
+            change_list.pop();
```

Comment:
Yeah because `new` and `old` are non-empty, we know we execute the loop at least once, and therefore push a child onto the stack, and that is what is being popped here. I can make a comment to this effect to clear things up.